File: src/test/java/com/ipseorama/sctp/SCTPMessageTest.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.ipseorama.sctp;

import com.ipseorama.sctp.messages.Chunk;
import com.ipseorama.sctp.messages.DataChunk;
import com.ipseorama.sctp.messages.SackChunk;
import com.ipseorama.sctp.messages.exceptions.SctpPacketFormatException;
import com.ipseorama.sctp.small.BlockingSCTPStream;
import java.io.IOException;
import java.util.TreeSet;
import org.bouncycastle.crypto.tls.DatagramTransport;
import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import static org.junit.Assert.*;

/**
 *
 * @author tim
 */
public class SCTPMessageTest {

    private SCTPStream _fakeStream;

    public SCTPMessageTest() {
    }

    @BeforeClass
    public static void setUpClass() {
    }

    @AfterClass
    public static void tearDownClass() {
    }
    Association _fakeAssociation;

    @Before
    public void setUp() {
        DatagramTransport _fakedt = new DatagramTransport() {

            @Override
            public int getReceiveLimit() throws IOException {
                return 1200;
            }

            @Override
            public int getSendLimit() throws IOException {
                return 1200;
            }

            @Override
            public int receive(byte[] bytes, int i, int i1, int waitMs) throws IOException {
                try {
                    Thread.sleep(waitMs);
                } catch (Exception x) {
                }
                throw new java.io.InterruptedIOException("empty");
            }

            @Override
            public void send(byte[] bytes, int i, int i1) throws IOException {
            }

            @Override
            public void close() throws IOException {
            }

        };
        _fakeAssociation = new Association(_fakedt, null) {
            @Override
            public void associate() throws SctpPacketFormatException, IOException {
                throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
            }

            @Override
            public void enqueue(DataChunk d) {
                throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
            }

            @Override
            public SCTPStream mkStream(int id) {
                throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
            }

            @Override
            public void sendAndBlock(SCTPMessage m) throws Exception {
                throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
            }

            @Override
            public SCTPMessage makeMessage(byte[] bytes, BlockingSCTPStream aThis) {
                throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
            }

            @Override
            public SCTPMessage makeMessage(String s, BlockingSCTPStream aThis) {
                throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
            }

            @Override
            protected Chunk[] sackDeal(SackChunk sackChunk) {
                throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
            }
        };
        _fakeStream = new SCTPStream(_fakeAssociation, new Integer(22)) {
            @Override
            public void send(String message) throws Exception {
                throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
            }

        };
    }

    @After
    public void tearDown() {
    }

    /**
     * Test of setCompleteHandler method, of class SCTPMessage.
     */
    /*
     @Test
     public void testSetCompleteHandler() {
     System.out.println("setCompleteHandler");
     MessageCompleteHandler mch = null;
     SCTPMessage instance = null;
     instance.setCompleteHandler(mch);
     // TODO review the generated test code and remove the default call to fail.
     fail("The test case is a prototype.");
     }
     */
    /**
     * Test of hasMoreData method, of class SCTPMessage.
     */
    /*
     @Test
     public void testHasMoreData() {

     }
     */
    /**
     * Test of fill method, of class SCTPMessage.
     */
    @Test
    public void testFillShortString() {
        System.out.println("--> fill short");
        String testString = "This is a short test";
        SCTPMessage instance = new SCTPMessage(testString, _fakeStream);
        TreeSet<DataChunk> chunks = new TreeSet<DataChunk>();
        while (instance.hasMoreData()) {
            DataChunk dc = new DataChunk();
            instance.fill(dc);
            chunks.add(dc);
        }
        assertEquals(chunks.size(), 1);
    }

    @Test
    public void testFillLongString() {
        System.out.println("--> fill long");
        StringBuffer sb = new StringBuffer("This is a");
        for (int i = 0; i < 1030; i++) {
            sb.append(" long");
        }
        sb.append(" test.");
        String testString = sb.toString();
        SCTPMessage instance = new SCTPMessage(testString, _fakeStream);
        TreeSet<DataChunk> chunks = new TreeSet<DataChunk>();
        long tsn = 111;

        while (instance.hasMoreData()) {
            DataChunk dc = new DataChunk();
            dc.setTsn(tsn++);
            instance.fill(dc);
            chunks.add(dc);
        }
        double pktsz = chunks.first().getDataSize();
        int estimate = (int) Math.ceil(testString.length() / pktsz);
        assertEquals(chunks.size(), estimate);
    }

    /**
     * Test of getData method, of class SCTPMessage.
     */
    /*
     @Test
     public void testGetData() {
     System.out.println("getData");
     SCTPMessage instance = null;
     byte[] expResult = null;
     byte[] result = instance.getData();
     assertArrayEquals(expResult, result);
     // TODO review the generated test code and remove the default call to fail.
     fail("The test case is a prototype.");
     }*/
}

File: tests/src/test/java/pe/pi/sctp4j/sctp/SCTPMessageTest.cs
/*
 * Copyright 2017 pi.pe gmbh .
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
// Modified by Andrés Leone Gámez


using Microsoft.VisualStudio.TestTools.UnitTesting;
using Org.BouncyCastle.Crypto.Tls;
using pe.pi.sctp4j.sctp.messages;
using pe.pi.sctp4j.sctp.small;
using SCTP4CS.Utils;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;

/**
*
* @author tim
*/
namespace pe.pi.sctp4j.sctp {
	[TestClass]
	public class SCTPMessageTest {

		private Random _rand = new Random(1); // deterministic non crypto quality random for repeatable tests
		private SCTPStream _fakeStream;

		public SCTPMessageTest() {
		}

		Association _fakeAssociation;

		class DatagramTransportImpl : DatagramTransport {
			public int GetReceiveLimit() {
				return 1200;
			}

			public int GetSendLimit() {
				return 1200;
			}

			public int Receive(byte[] bytes, int i, int i1, int waitMs) {
				try {
					Thread.Sleep(waitMs);
				}
				catch (Exception x) {
				}
				throw new Exception("[InterruptedIOException] empty");
			}

			public void Send(byte[] bytes, int i, int i1) { }
			public void Close() { }
		}
		class AssociationImpl : Association {
			public AssociationImpl(DatagramTransport transport, AssociationListener al) : base(transport, al) {
			}

			public AssociationImpl(DatagramTransport transport, AssociationListener al, bool client) : base(transport, al, client) {
			}

			public override void associate() {
				throw new Exception("[UnsupportedOperationException] Not supported yet."); //To change body of generated methods, choose Tools | Templates.
			}

			internal override void enqueue(DataChunk d) {
				throw new Exception("[UnsupportedOperationException] Not supported yet."); //To change body of generated methods, choose Tools | Templates.
			}

			public override SCTPStream mkStream(int id) {
				throw new Exception("[UnsupportedOperationException] Not supported yet."); //To change body of generated methods, choose Tools | Templates.
			}

			internal override void sendAndBlock(SCTPMessage m) {
				throw new Exception("[UnsupportedOperationException] Not supported yet."); //To change body of generated methods, choose Tools | Templates.
			}

			internal override SCTPMessage makeMessage(byte[] bytes, int offset, int length, BlockingSCTPStream aThis) {
				throw new Exception("[UnsupportedOperationException] Not supported yet."); //To change body of generated methods, choose Tools | Templates.
			}

			internal override SCTPMessage makeMessage(string s, BlockingSCTPStream aThis) {
				throw new Exception("[UnsupportedOperationException] Not supported yet."); //To change body of generated methods, choose Tools | Templates.
			}

			internal override Chunk[] sackDeal(SackChunk sackChunk) {
				throw new Exception("[UnsupportedOperationException] Not supported yet."); //To change body of generated methods, choose Tools | Templates.
			}
		}
		class SCTPStreamImpl : SCTPStream {
			public SCTPStreamImpl(Association a, int id) : base(a, id) {
			}

			public override void send(string message) {
				throw new Exception("[UnsupportedOperationException] Not supported yet."); //To change body of generated methods, choose Tools | Templates.
			}

			internal override void deliverMessage(SCTPMessage message) {
				message.run();
			}

			public override void send(byte[] message, int offset, int length) {
				throw new Exception("[UnsupportedOperationException] Not supported yet."); //To change body of generated methods, choose Tools | Templates.
			}

			internal override void delivered(DataChunk d) {
				throw new Exception("[UnsupportedOperationException] Not supported yet."); //To change body of generated methods, choose Tools | Templates.
			}
		}

		[TestInitialize]
		public void setUp() {
			DatagramTransport _fakedt = new DatagramTransportImpl();
			_fakeAssociation = new AssociationImpl(_fakedt, null);
			_fakeStream = new SCTPStreamImpl(_fakeAssociation, 22);
		}

		/**
		 * Test of setCompleteHandler method, of class SCTPMessage.
		 */
		/*
		 @Test
		 public void testSetCompleteHandler() {
		 Console.WriteLine("setCompleteHandler");
		 MessageCompleteHandler mch = null;
		 SCTPMessage instance = null;
		 instance.setCompleteHandler(mch);
		 // TODO review the generated test code and remove the default call to fail.
		 fail("The test case is a prototype.");
		 }
		 */
		/**
		 * Test of hasMoreData method, of class SCTPMessage.
		 */
		/*
		 @Test
		 public void testHasMoreData() {

		 }
		 */
		/**
		 * Test of fill method, of class SCTPMessage.
		 */
		[TestMethod]
		public void testFillShortstring() {
			Console.WriteLine("--> fill short string ");
			string teststring = "This is a short test";
			SCTPMessage instance = new SCTPMessage(teststring, _fakeStream);
			SortedArray<DataChunk> chunks = new SortedArray<DataChunk>();
			while (instance.hasMoreData()) {
				DataChunk dc = new DataChunk();
				instance.fill(dc);
				chunks.Add(dc);
			}
			Assert.AreEqual(chunks.Count, 1);
			var ppid = ((DataChunk) chunks.First).ppid;
			Assert.AreEqual(ppid, SCTP_PPID.WEBRTCstring);
		}
		[TestMethod]
		public void testFillShortBlob() {
			Console.WriteLine("--> fill short blob ");
			byte[] testBlob = new byte[21];
			_rand.NextBytes(testBlob);
			SCTPMessage instance = new SCTPMessage(testBlob, 0, testBlob.Length, _fakeStream);
			SortedArray<DataChunk> chunks = new SortedArray<DataChunk>();
			while (instance.hasMoreData()) {
				DataChunk dc = new DataChunk();
				instance.fill(dc);
				chunks.Add(dc);
			}
			Assert.AreEqual(chunks.Count, 1);
			var ppid = ((DataChunk) chunks.First).ppid;
			Assert.AreEqual(ppid, SCTP_PPID.WEBRTCBINARY);
		}
		[TestMethod]
		public void testFillLongstring() {
			Console.WriteLine("--> fill long");
			StringBuilder sb = new StringBuilder("This is a");
			for (int i = 0; i < 1030; i++) {
				sb.Append(" long");
			}
			sb.Append(" test.");
			string teststring = sb.ToString();
			SCTPMessage instance = new SCTPMessage(teststring, _fakeStream);
			SortedArray<DataChunk> chunks = new SortedArray<DataChunk>();
			long tsn = 111;

			while (instance.hasMoreData()) {
				DataChunk dc = new DataChunk();
				dc.tsn = (uint) tsn++;
				instance.fill(dc);
				chunks.Add(dc);
			}
			double pktsz = chunks.First.getDataSize();
			int estimate = (int) Math.Ceiling(teststring.Length / pktsz);
			Assert.AreEqual(chunks.Count, estimate);
		}

		[TestMethod]
		public void testEmptystring() {
			Console.WriteLine("--> fill empty string");
			StringBuilder sb = new StringBuilder("");
			string teststring = sb.ToString();
			SCTPMessage instance = new SCTPMessage(teststring, _fakeStream);
			SortedArray<DataChunk> chunks = new SortedArray<DataChunk>();
			long tsn = 111;

			while (instance.hasMoreData()) {
				DataChunk dc = new DataChunk();
				dc.tsn = (uint) tsn++;
				instance.fill(dc);
				chunks.Add(dc);
			}
			int pktsz = chunks.First.getDataSize();
			Assert.AreEqual(chunks.Count, 1);
			Assert.AreEqual(pktsz, 1);
			var ppid = ((DataChunk) chunks.First).ppid;
			Assert.AreEqual(ppid, SCTP_PPID.WEBRTCstringEMPTY);
		}

		[TestMethod]
		public void testEmptyBlob() {
			Console.WriteLine("--> fill empty blob");
			byte[] testBlob = new byte[0];
			SCTPMessage instance = new SCTPMessage(testBlob, 0, testBlob.Length, _fakeStream);
			SortedArray<DataChunk> chunks = new SortedArray<DataChunk>();
			long tsn = 111;

			while (instance.hasMoreData()) {
				DataChunk dc = new DataChunk();
				dc.tsn = (uint) tsn++;
				instance.fill(dc);
				chunks.Add(dc);
			}
			Assert.AreEqual(chunks.Count, 1);
			int pktsz = chunks.First.getDataSize();
			Assert.AreEqual(pktsz, 1);
			var ppid = ((DataChunk) chunks.First).ppid;
			Assert.AreEqual(ppid, SCTP_PPID.WEBRTCBINARYEMPTY);
		}

		/**
		 * Test of getData method, of class SCTPMessage.
		 */
		/*
		 @Test
		 public void testGetData() {
		 Console.WriteLine("getData");
		 SCTPMessage instance = null;
		 byte[] expResult = null;
		 byte[] result = instance.getData();
		 assertArrayEquals(expResult, result);
		 // TODO review the generated test code and remove the default call to fail.
		 fail("The test case is a prototype.");
		 }*/
	}
}
